pmda/density.py

```diff
@@ -0,0 +1,144 @@
+    """
+    def __init__(self, atomgroup, delta=1.0, atomselection="name OH2",
+                start=None, stop=None, step=None, metadata=None, padding=2.0,
+                update_selection=False, parameters=None, gridcenter=None,
+                xdim=None, ydim=None, zdim=None):
+        u = atomgroup.universe
+        super(DensityAnalysis, self).__init__(u, (atomgroup, ))
+        self._atomgroup = atomgroup
+        self._delta = delta
+        self._atomselection = atomselection
+        self._metadata = metadata
+        self._padding = padding
+        self._update_selection = update_selection
+        self._parameters = parameters
+        self._gridcenter = gridcenter
+        self._xdim = xdim
+        self._ydim = ydim
+        self._zdim = zdim
+        self._trajectory = u.trajectory
+        self._n_frames = u.trajectory.n_frames
+
+    def _prepare(self):
+        coord = self.current_coordinates(self._atomgroup, self._atomselection,
+        self._update_selection)
+        box, angles = self._trajectory.ts.dimensions[:3],
+        self._trajectory.ts.dimensions[3:]
+        if self._gridcenter is not None:
+            # Generate a copy of smin/smax from coords to later check if the
+            # defined box might be too small for the selection
+            smin = np.min(coord, axis=0)
+            smax = np.max(coord, axis=0)
+            # Overwrite smin/smax with user defined values
+            smin, smax = _set_user_grid(self._gridcenter, self._xdim,
+            self._ydim, self._zdim, smin, smax)
+        else:
+            # Make the box bigger to avoid as much as possible 'outlier'. This
+            # is important if the sites are defined at a high density: in this
+            # case the bulk regions don't have to be close to 1 * n0 but can
+            # be less. It's much more difficult to deal with outliers.  The
+            # ideal solution would use images: implement 'looking across the
+            # periodic boundaries' but that gets complicated when the box
+            # rotates due to RMS fitting.
+            smin = np.min(coord, axis=0) - self._padding
+            smax = np.max(coord, axis=0) + self._padding
+            BINS = fixedwidth_bins(self._delta, smin, smax)
+            arange = np.transpose(np.vstack((BINS['min'], BINS['max'])))
+            bins = BINS['Nbins']
+            # create empty grid with the right dimensions (and get the edges)
+            grid, edges = np.histogramdd(np.zeros((1, 3)), bins=bins,
+            range=arange, normed=False)
+            grid *= 0.0
+        self._grid = grid
+        self._edges = edges
+        self._arange = arange
+        self._bins = bins
+
+    def _single_frame(self, ts, atomgroups):
+        coord = self.current_coordinates(atomgroups[0], self._atomselection,
+        self._update_selection)
+        h, edges = np.histogramdd(coord, bins=self._bins, range=self._arange,
+        normed=False)
+        return [np.array(h), edges]
```

Comment:
You shouldn't need to `np.array(h)` – isn't it already an array? Just do
```python
return h, edges
```

Reply:
I tried running it as a tuple but it produced several issues in the reduce function due to the indexing. Removing the indexing caused a run-time error that filled 32 GB of memory in a few seconds. It never reached the conclude function, which would also need to be fixed due to indexing. This change will require further inspection of ParallelAnalysisBase before it can be made.

Reply:
Can you paste the error?

Reply:
Also, looking at the code, you're not using `edges` in `_reduce`: just only 
```python
return h
```
and adapt your `_reduce` function.